pennylane/interfaces/batch/jax.py

```diff
@@ -1,299 +1,351 @@
-
-    def wrapped_exec_bwd(params, g):
-
-        if isinstance(gradient_fn, qml.gradients.gradient_transform):
-
-            def non_diff_wrapper(args):
-                """Compute the VJP in a non-differentiable manner."""
-                new_tapes = []
-                p = args[:-1]
-                dy = args[-1]
-
-                for t, a in zip(tapes, p):
-                    new_tapes.append(t.copy(copy_operations=True))
-                    new_tapes[-1].set_parameters(a)
-                    new_tapes[-1].trainable_params = t.trainable_params
-
-                vjp_tapes, processing_fn = qml.gradients.batch_vjp(
-                    new_tapes,
-                    dy,
-                    gradient_fn,
-                    reduction="append",
-                    gradient_kwargs=gradient_kwargs,
-                )
-
-                partial_res = execute_fn(vjp_tapes)[0]
-                res = processing_fn(partial_res)
-                return np.concatenate(res)
-
-            args = tuple(params) + (g,)
-            vjps = host_callback.call(
-                non_diff_wrapper,
-                args,
-                result_shape=jax.ShapeDtypeStruct((total_params,), dtype),
-            )
-
-            param_idx = 0
-            res = []
-
-            # Group the vjps based on the parameters of the tapes
-            for p in params:
-                param_vjp = vjps[param_idx : param_idx + len(p)]
-                res.append(param_vjp)
-                param_idx += len(p)
-
-            # Unwrap partial results into ndim=0 arrays to allow
-            # differentiability with JAX
-            # E.g.,
-            # [DeviceArray([-0.9553365], dtype=float32), DeviceArray([0., 0.],
-            # dtype=float32)]
-            # is mapped to
-            # [[DeviceArray(-0.9553365, dtype=float32)], [DeviceArray(0.,
-            # dtype=float32), DeviceArray(0., dtype=float32)]].
-            need_unwrapping = any(r.ndim != 0 for r in res)
-            if need_unwrapping:
-                unwrapped_res = []
-                for r in res:
-                    if r.ndim != 0:
-                        r = [jnp.array(p) for p in r]
-                    unwrapped_res.append(r)
-
-                res = unwrapped_res
-
-            return (tuple(res),)
-
-        # Gradient function is a device method.
-        with qml.tape.Unwrap(*tapes):
-            jacs = gradient_fn(tapes, **gradient_kwargs)
-
-        vjps = [qml.gradients.compute_vjp(d, jac) for d, jac in zip(g, jacs)]
-        res = [[jnp.array(p) for p in v] for v in vjps]
-        return (tuple(res),)
-
-    wrapped_exec.defvjp(wrapped_exec_fwd, wrapped_exec_bwd)
-    return wrapped_exec(params)
-
-
-# The execute function in forward mode
-def _execute_with_fwd(
-    params,
-    tapes=None,
-    device=None,
-    execute_fn=None,
-    gradient_kwargs=None,
-    _n=1,
-):  # pylint: disable=dangerous-default-value,unused-argument
-
-    _validate_tapes(tapes)

```

Comment:
The content of this old implementation were moved to `jax_jit.py` but the `_validate_tapes(tapes)` calls in the `_execute()` and `_execute_with_fwd()` were removed. Why was this done? It seems that a call was added to `execute()` instead, is this a case where instead of checking in the private function calls you moved the validation check to the base call?

Reply:
Yep exactly, just simplified the code to have the same validation be done earlier in the pipeline.